cmd/sync/aws.go

```diff
@@ -152,17 +166,61 @@ func (client *AWSClient) GetPrivateIPsForScalingGroup(name string) ([]string, er
 	}
 
 	var result []string
+	insIDtoIP := make(map[string]string)
+
 	for _, res := range response.Reservations {
 		for _, ins := range res.Instances {
 			if len(ins.NetworkInterfaces) > 0 && ins.NetworkInterfaces[0].PrivateIpAddress != nil {
-				result = append(result, *ins.NetworkInterfaces[0].PrivateIpAddress)
+				if onlyInService {
+					insIDtoIP[*ins.InstanceId] = *ins.NetworkInterfaces[0].PrivateIpAddress
+				} else {
+					result = append(result, *ins.NetworkInterfaces[0].PrivateIpAddress)
+				}
 			}
 		}
 	}
+	if onlyInService {
+		result = client.getInstancesInService(insIDtoIP)
+	}
 
 	return result, nil
 }
 
+// getInstancesInService returns the list of instances that have LifecycleState == InService
+func (client *AWSClient) getInstancesInService(insIDtoIP map[string]string) []string {
+	const maxItems = 50
+	var result []string
+	keys := reflect.ValueOf(insIDtoIP).MapKeys()
+	instanceIds := make([]*string, len(keys))
+
+	for i := 0; i < len(keys); i++ {
+		instanceIds[i] = aws.String(keys[i].String())
+	}
+
+	min := func(a, b int) int {
+		if a <= b {
+			return a
+		}
+		return b
+	}
+
+	for i := 0; i < len(instanceIds); i += maxItems {
+		// the API only accepts 50 ids at once
+		batch := instanceIds[i:min(i+maxItems, len(instanceIds))]
```

Comment:
I wonder if it makes sense to have a unit test for this logic?
what if you have a function like `prepareBatches()` that returns a `[][]*string` ? that function could be unit tested